### ISSUE TITLE
Jetty spans disappear during process execution

### ISSUE THREAD

**Issue by michaeljohnalbers** [open-telemetry/opentelemetry-java-instrumentation#8415]
**Describe the bug**

On April 11th, Jetty spans (as well as AWS SDK v2) spans mysteriously started disappearing after our Java process started. Upon startup the number of spans appears normal-ish, but over the next 10-20 minutes the number of spans drops off to almost nothing even though the process is still responding to HTTP requests.

In this screenshot (from Honeycomb.io, where we send our spans) you can see the massive drop-off. Everything before April 11th is nominal. Everything after is not.

![image](https://user-images.githubusercontent.com/8935186/236293545-5ea9316b-859a-47c3-914a-042f80a963de.png)

In this image you can see a close up view of what happens after a deployment of our process.

![image](https://user-images.githubusercontent.com/8935186/236293781-0e7e10de-3fa4-4c78-a82e-fafbcd8d9051.png)

I've taken many steps to attempt to fix the issue including:
* Reverting numerous changes deployed around the 11th
* Reverting a few dependency updates which happened around the 11th
* Upgrading the OTel agent
* Disabling Jetty instrumentation
* Turning off trace sampling (we normally sample 1:8 traces)

Obviously, none of these have fixed the problem.

We have a mix of automatic instrumentation and manual instrumentation in our process. These worked fine with each other before the 11th. The manually instrumented traces continue to be reported in normal amounts.

I've worked with the Honeycomb team on this. They've verified the problem is not on their end.

Looking at the data a little closer shows that what should be child spans of the Jetty spans are all being accumulated under very few trace IDs. Normally a HTTP request results in around 10-30 spans in total. I'm seeing traces with well over 1,000,000 spans, which are obviously for multiple separate HTTP requests. The way our system is architected, each HTTP request should be the start of a new trace.

I'm at wits end to what might be causing this and how to fix it. I'm not asking anyone here to fix the problem, but I'd love some guidance on what might potentially be causing this (any ideas are welcome) or what steps I might take to debug/troubleshoot the issue.

**Steps to reproduce**
Just running our application makes it happen. Beyond that I have no idea.

**What did you expect to see?**
Far, far more Jetty spans.

**What did you see instead?**
Small spikes of spans after restarts, then tailing off to nearly none.

**What version are you using?**
1.22.1 and 1.25.0
Honeycomb versions 1.4.3 and 1.5.0
Currently running 1.25.0/1.5.0

**Environment**
Compiler: OpenJDK 11 (from Circle CI docker image `cimg/openjdk:11.0`)
OS: Ubuntu 22.04
Runtime (if different from JDK above): GraalVM 22.3.1/JDK 17
OS (if different from OS compiled on): Ubunut 22.10

**Additional context**
Our process is a DropWizard app (v2.1.6) with Jetty 9.4.51. We're running in Docker containers within AWS ECS.

The DropWizard/Jetty versions, as well as the compiler/runtime Java versions, did not change anywhere near the start of this problem.

**Comment by trask** [comment]
> Looking at the data a little closer shows that what should be child spans of the Jetty spans are all being accumulated under very few trace IDs. Normally a HTTP request results in around 10-30 spans in total. I'm seeing traces with well over 1,000,000 spans, which are obviously for multiple separate HTTP requests.

this sounds like a context leak, where the context from a prior request is left attached to the jetty thread when it is returned back to the thread pool

check out:

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/413890d246253f65831dd07bc51a38e7b19cedd8/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/PropagatingFromUpstreamInstrumenter.java#L28

this is where the jetty SERVER span is created, and you will probably see that there is already an active Context (with another span in it already)

if you follow that call to `ContextPropagationDebug.debugContextLeakIfEnabled()` there are debug options that you can enable that will hopefully give you more info about what's going on and where the leak is coming from

**Comment by trask** [comment]
since you mentioned you have manual instrumentation as well, it's worth checking that you are always using `.makeCurrent()` inside of a try-with-resources block, since not closing Scope is an easy way to leak context

**Comment by michaeljohnalbers** [comment]
I verified that a few days ago.

**Comment by michaeljohnalbers** [comment]
@trask I enabled the output messages in ContextPropagationDebug by setting the `OTEL_JAVAAGENT_EXPERIMENTAL_THREAD_PROPAGATION_DEBUGGER_ENABLED` env var to `true`. I've seen none of the debug output from that class.

**Comment by laurit** [comment]
@michaeljohnalbers have you tried building a minimal application that reproduces the issue? When you look at the traces where all the spans are accumulating do they all start with the same endpoint? If they do then probably the implementation of that endpoint does something do provoke the issue and could help reproducing it.

**Comment by michaeljohnalbers** [comment]
@laurit I have not tried to create a minimal app to reproduce the issue. There are complications with that, but it is worth thinking about.

The traces which accumulate spans in the millions are all missing the root span. I've looked in one other Honeycomb dataset where it might have existed but didn't come up with anything. However, I do see a few traces which don't have millions of child spans, but do have thousands (which is just as incorrect). And in those two cases it is the same endpoint. Thanks for that suggestion, I'll take a further look.

**Comment by michaeljohnalbers** [comment]
I turned on full OTel debug logging (`-Dotel.javaagent.debug=true`) in our application and found a few things worth noting.

I can see the output from `io.opentelemetry.exporter.logging.LoggingSpanExporter` drop off to nothing for `tracer: io.opentelemetry.jetty-8.0:1.25.0-alpha`. Presumably it means that this isn't a problem in sending data to Honeycomb (though I figured that was the case, at least this should confirm it) and that the internal state is getting messed up. The spans for `io.opentelemetry.aws-sdk-2.2:1.25.0-alpha` drop off in pretty much the same timeframe. However, after the Jetty spans stop I am still seeing the sporadic single AWS span as the process continues to run.

I filtered out all the log messages from `io.opentelemetry.exporter.logging.LoggingSpanExporter` as well as our logging instances and there are zero log messages after the application starts up.

During startup there are some jaxrs warnings that look fishy.
> Instrumentation skipped, mismatched references were found: jaxrs [class io.opentelemetry.javaagent.instrumentation.jaxrs.v3_0.JerseyInstrumentationModule] on org.springframework.boot.loader.LaunchedURLClassLoader@730f9695
-- io.opentelemetry.javaagent.instrumentation.jaxrs.v3_0.Jaxrs3HandlerData:57 Missing class jakarta.ws.rs.Path
-- io.opentelemetry.javaagent.instrumentation.jaxrs.v3_0.JerseyResourceMethodDispatcherInstrumentation$DispatchAdvice:44 Missing class jakarta.ws.rs.core.Request
-- io.opentelemetry.javaagent.instrumentation.jaxrs.v3_0.Jaxrs3HandlerData:52 Missing class jakarta.ws.rs.HttpMethod
-- io.opentelemetry.javaagent.instrumentation.jaxrs.v3_0.JerseyRequestContextInstrumentation$ContainerRequestContextAdvice:44 Missing class jakarta.ws.rs.core.UriInfo
-- io.opentelemetry.javaagent.instrumentation.jaxrs.v3_0.JerseyRequestContextInstrumentation$ContainerRequestContextAdvice:46 Missing class jakarta.ws.rs.container.ResourceInfo
-- io.opentelemetry.javaagent.instrumentation.jaxrs.v3_0.JerseyRequestContextInstrumentation$ContainerRequestContextAdvice:44 Missing class jakarta.ws.rs.container.ContainerRequestContext
-- io.opentelemetry.javaagent.instrumentation.jaxrs.v3_0.JerseyServletContainerInstrumentation$ServiceAdvice:49 Missing class jakarta.servlet.http.HttpServletRequest
>
>Instrumentation skipped, mismatched references were found: jaxrs [class io.opentelemetry.javaagent.instrumentation.jaxrs.v2_0.Resteasy30InstrumentationModule] on org.springframework.boot.loader.LaunchedURLClassLoader@730f9695
-- io.opentelemetry.javaagent.instrumentation.jaxrs.v2_0.ResteasyRootNodeTypeInstrumentation$AddInvokerAdvice:52 Missing class org.jboss.resteasy.core.ResourceLocatorInvoker
-- io.opentelemetry.javaagent.instrumentation.jaxrs.v2_0.Resteasy30RequestContextInstrumentation$ContainerRequestContextAdvice:45 Missing class org.jboss.resteasy.core.interception.PostMatchContainerRequestContext
-- io.opentelemetry.javaagent.instrumentation.jaxrs.v2_0.Resteasy31RequestContextInstrumentation$ContainerRequestContextAdvice:51 Missing class org.jboss.resteasy.core.ResourceMethodInvoker
-- io.opentelemetry.javaagent.instrumentation.jaxrs.v2_0.Resteasy31RequestContextInstrumentation$ContainerRequestContextAdvice:45 Missing class org.jboss.resteasy.core.interception.jaxrs.PostMatchContainerRequestContext
-- io.opentelemetry.javaagent.instrumentation.jaxrs.v2_0.ResteasyRootNodeTypeInstrumentation$AddInvokerAdvice:52 Missing class org.jboss.resteasy.core.ResourceLocatorInvoker

What do these mean? Are they normal/expected?

**Comment by laurit** [comment]
> During startup there are some jaxrs warnings that look fishy.

Before instrumentation is applied we check whether all classes used in the instrumentation are present to avoid runtime failures. These warnings say that some instrumentations weren't applied because they depend on classes that are not present. You can ignore these, they won't cause issues. To avoid these warnings we need add extra checks to instrumentation to verify that expected version of jax-rs api is present and that some framework specific class is present to avoid attempting to run resteasy instrumentation when jersey is used and vice versa.

As @trask said most likely the issue is that somehow context has leaked from a previous request and `SERVER` span created by jetty gets suppressed because there already is an `SERVER` span in the context. `ContextPropagationDebug` won't show this because span suppression logic will prevent the code from reaching there. You could try setting `-Dotel.instrumentation.experimental.span-suppression-strategy=none` which will probably get the jetty `SERVER` spans back. If you have a debug log from a run where this issue happens you should see the suppressed count in lines like https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/389005ea11c2244593a526284c520c74e2429c89/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/internal/SupportabilityMetrics.java#L71 this for jetty `SERVER` span increase. Hopefully disabling span suppression will let you get the debug info that `ContextPropagationDebug` prints out, this could allow us to figure out where the context leaks from.

**Comment by trask** [comment]
@michaeljohnalbers can you try out @laurit's fix? [a SNAPSHOT build](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/CONTRIBUTING.md#snapshot-builds) should be available once [the `main` build is complete](https://github.com/open-telemetry/opentelemetry-java-instrumentation/actions/runs/5061393031)

**Comment by michaeljohnalbers** [comment]
@trask Sure. Is this just a change to the agent or do I need to update the Maven dependency we use as well?

**Comment by trask** [comment]
just update the agent only

**Comment by michaeljohnalbers** [comment]
@trask @laurit That change seems to have fixed some of the problem. We're seeing more Jetty spans now. Nowhere near the level we should be, but it's a nice start. They are no longer tailing off to zero spans anymore.

With the 1.27 SNAPSHOT agent, I'm still seeing lots of
>Unexpected non-root current context found when extracting remote context!

log messages. In the "It contains this span:" message, the span information appears to always be for the name `s3.GetObject`. Our code never calls GetObject directly. We're using the [S3TransferManager](https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/transfer/s3/S3TransferManager.html).

Also I'm not seeing any "a context leak was detected" messages.

**Comment by josecoelho96** [comment]
Currently using the 1.27.0 version and this problem also happens to me.

Jetty version: 9.4.31.v20200723
Java Version (Runtime): 11.0.14

Only using automatic instrumentation, so the http request is the parent span. I know some endpoints are being called but only some are correctly visualized, while others "disappear". Usually my trace has 1 parent span (jetty) and one (or more) database related spans.

I'm using Tempo as a trace backend. If I navigate a bit there, sometimes I can find extremely long traces (several minutes), composed of multiple spans, related to multiple http requests that have nothing to do with each other.

(I'm not sure if I should leave a comment here or create a new issue, so let me know if I should open a new issue)

**Comment by laurit** [comment]
@josecoelho96 please create a new issue. If possible include a sample app that reproduces your issue.
